counterpartylib/lib/transaction.py

```diff
@@ -340,39 +313,63 @@ def construct (db, tx_info, encoding='auto',
 
     source_is_p2sh = script.is_p2sh(source)
 
+    # Normalize source
+    if script.is_multisig(source):
+        source_address = backend.multisig_pubkeyhashes_to_pubkeys(source, provided_pubkeys)
+    else:
+        source_address = source
+
     # Sanity checks.
     if exact_fee and not isinstance(exact_fee, int):
         raise exceptions.TransactionError('Exact fees must be in satoshis.')
     if not isinstance(fee_provided, int):
         raise exceptions.TransactionError('Fee provided must be in satoshis.')
 
-    if UTXO_LOCKS is None and config.UTXO_LOCKS_MAX_ADDRESSES > 0:  # initialize if configured
-        UTXO_LOCKS = util.DictCache(size=config.UTXO_LOCKS_MAX_ADDRESSES)
+    '''Determine encoding method'''
+
+    if data:
+        desired_encoding = encoding
+        # Data encoding methods (choose and validate).
+        if desired_encoding == 'auto':
+            if len(data) + len(config.PREFIX) <= config.OP_RETURN_MAX_SIZE:
+                encoding = 'opreturn'
+            else:
+                encoding = 'p2sh' if not old_style_api and util.enabled('p2sh_encoding') else 'multisig'  # p2sh is not possible with old_style_api
```

Comment:
should we also check `and not (source_is_p2sh and dust_return_pubkey is None)`? because if that's not true then we'd fail later